lib/grant-types/refresh-token-grant-type.ts

```diff
@@ -0,0 +1,158 @@
+      throw new InvalidArgumentError(
+        'Invalid argument: model does not implement `revokeToken()`',
+      );
+    }
+
+    if (!options.model.saveToken) {
+      throw new InvalidArgumentError(
+        'Invalid argument: model does not implement `saveToken()`',
+      );
+    }
+  }
+
+  /**
+   * Handle refresh token grant.
+   *
+   * @see https://tools.ietf.org/html/rfc6749#section-6
+   */
+
+  async handle(request: Request, client: Client) {
+    if (!request) {
+      throw new InvalidArgumentError('Missing parameter: `request`');
+    }
+
+    if (!client) {
+      throw new InvalidArgumentError('Missing parameter: `client`');
+    }
+
+    const token = await this.getRefreshToken(request, client);
+    await this.revokeToken(token);
+
+    return this.saveToken(token.user, client, token.scope);
+  }
+
+  /**
+   * Get refresh token.
+   */
+
+  async getRefreshToken(request: Request, client: Client) {
+    if (!request.body.refresh_token) {
+      throw new InvalidRequestError('Missing parameter: `refresh_token`');
+    }
+
+    if (!is.vschar(request.body.refresh_token)) {
+      throw new InvalidRequestError('Invalid parameter: `refresh_token`');
+    }
+
+    const token = await this.model.getRefreshToken(request.body.refresh_token);
+
+    if (!token) {
+      throw new InvalidGrantError('Invalid grant: refresh token is invalid');
+    }
+
+    if (!token.client) {
+      throw new ServerError(
+        'Server error: `getRefreshToken()` did not return a `client` object',
+      );
+    }
+
+    if (!token.user) {
+      throw new ServerError(
+        'Server error: `getRefreshToken()` did not return a `user` object',
+      );
+    }
+
+    if (token.client.id !== client.id) {
+      throw new InvalidGrantError('Invalid grant: refresh token is invalid');
+    }
+
+    if (
+      token.refreshTokenExpiresAt &&
+      !(token.refreshTokenExpiresAt instanceof Date)
+    ) {
+      throw new ServerError(
+        'Server error: `refreshTokenExpiresAt` must be a Date instance',
+      );
+    }
+
+    if (
+      token.refreshTokenExpiresAt &&
+      token.refreshTokenExpiresAt.getTime() < Date.now()
+    ) {
+      throw new InvalidGrantError('Invalid grant: refresh token has expired');
+    }
+
+    return token;
+  }
+
+  /**
+   * Revoke the refresh token.
+   *
+   * @see https://tools.ietf.org/html/rfc6749#section-6
+   */
+
+  async revokeToken(token: RefreshToken) {
+    if (this.alwaysIssueNewRefreshToken === false) {
+      return token;
+    }
+
+    const status = await this.model.revokeToken(token);
+    if (!status) {
+      throw new InvalidGrantError('Invalid grant: refresh token is invalid');
+    }
+
+    return token;
+  }
+
+  /**
+   * Save token.
+   */
+
+  async saveToken(user: User, client: Client, scope: string) {
+    const accessToken = await this.generateAccessToken(client, user, scope);
+    const refreshToken = await this.generateRefreshToken(client, user, scope);
```

Comment:
These might resolve faster in parallel.

Reply:
Like I commented above

In this case it is better to not run them all in parallel just because it will process faster as it is security relevant and we need to validate the scope before we do the next steps